### ISSUE TITLE
ES 5 Doesn't allow registering default analyzer

### ISSUE THREAD

**Issue by mitchswaldman** [elastic/elasticsearch#19163]
**Elasticsearch version**: 5.0.0-alpha3

**JVM version**: 1.8.0_73-b02

**OS version**: CentOS Linux release 7.2.1511

**Description of the problem including expected versus actual behavior**:
I'm trying to upgrade an ES plugin from 2.3 to 5.0.0-alpha3, and a template we previously had working is now causing an issue. The template attempts to set some default analyzers for an index, but when the index is created it throws an error saying there's already an analyzer with the default name. 

Here's the template I'm using:
`{
  "order": 0, 
  "settings": {
    "index.analysis.analyzer.fairhairIndexAnalyzerv3.alias": "default",
    "index.analysis.analyzer.fairhairIndexAnalyzerv3.type": "fairhair-index-analyzer",
    "index.analysis.analyzer.fairhairTokenizingAnalyzer.type": "fairhair-tokenizing-analyzer",
    "index.analysis.analyzer.fairhairTokenizingAnalyzer.alias": "default_search"
  },
"mappings" : {
   //some mappings
},
"template" : "document.*"
}`

**Provide logs (if relevant)**:
Here's the error stack trace I get:
`elasticsearch_1  | [2016-06-28 23:40:43,871][DEBUG][action.admin.indices.create] [Iridia] [document.document-20151106] failed to create
elasticsearch_1  | java.lang.IllegalStateException: already registered analyzer with name: default
elasticsearch_1  | 	at org.elasticsearch.index.analysis.AnalysisService.<init>(AnalysisService.java:109)
elasticsearch_1  | 	at org.elasticsearch.index.analysis.AnalysisRegistry.build(AnalysisRegistry.java:161)
elasticsearch_1  | 	at org.elasticsearch.index.IndexService.<init>(IndexService.java:138)
elasticsearch_1  | 	at org.elasticsearch.index.IndexModule.newIndexService(IndexModule.java:328)
elasticsearch_1  | 	at org.elasticsearch.indices.IndicesService.createIndexService(IndicesService.java:398)
elasticsearch_1  | 	at org.elasticsearch.indices.IndicesService.createIndex(IndicesService.java:363)
`

I've looked at the AnalysisService class constructor, and I see it does a check for analyzers with the name "default" in the passed in analyzer mappings. I believe these conditional statements fire:

`if (!analyzerProviders.containsKey("default")) {
            analyzerProviders.put("default", new StandardAnalyzerProvider(indexSettings, null, "default", Settings.Builder.EMPTY_SETTINGS));
        }
        if (!analyzerProviders.containsKey("default_search")) {
            analyzerProviders.put("default_search", analyzerProviders.get("default"));
        }
        if (!analyzerProviders.containsKey("default_search_quoted")) {
            analyzerProviders.put("default_search_quoted", analyzerProviders.get("default_search"));
        }`

So it registers the standard analyzer, but while iterating through the passed in analyzers, this check also fires:

`if (analyzers.containsKey(name)) { //Grabs the name from each entry in analyzer map
                throw new IllegalStateException("already registered analyzer with name: " + name);
            }`

I'm confused as to how both these conditions are firing. This check to see if the current map of analyzers already contains the key "name" was introduced in ES 5. 

EDIT: I know why the illegal state exception is being thrown. The Analysis Registry automatically adds the entry: "default" => StandardAnalyzer. So from what I can tell, it's actually impossible to register your own default index analyzer, since there will always be the standard analyzer. I think the check for whether a name is already in the analyzer mappings is a bug and should be removed.

**Comment by s1monw** [comment]
that is a bug - I think we should not register a "default" analyzer in our `AnalysisModule` since `AnalysisService` does it for us if there is no default specified.

**Comment by mitchswaldman** [comment]
I think throwing an illegal state exception when encountering a duplicate analyzer name should be removed. If user defined analyzers are deterministically placed into the analyzer mappings after the built in analyzers, then I think it makes sense to simply overwrite them. Maybe I'm not clear on how all this stuff get's wired up, but I'm actually a little confused as to how it's even possible to trigger duplicate analyzer named form within the constructor of AnalysisService. A Map of (analyzer name => analyzer) is passed into the constructor, which by definition has no duplicate keys. So how is it that while iterating through this map that we even encounter the same key twice?

**Comment by s1monw** [comment]
> I think throwing an illegal state exception when encountering a duplicate analyzer name should be removed. If user defined analyzers are deterministically placed into the analyzer mappings after the built in analyzers, then I think it makes sense to simply overwrite them. Maybe I'm not clear on how all this stuff get's wired up, but I'm actually a little confused as to how it's even possible to trigger duplicate analyzer named form within the constructor of AnalysisService. A Map of (analyzer name => analyzer) is passed into the constructor, which by definition has no duplicate keys. So how is it that while iterating through this map that we even encounter the same key twice?

-1 default is a special one that's fine but nobody should override `english` analyzer etc. ever, you can define your own that's fine

**Comment by mitchswaldman** [comment]
True. Could you enlighten me on how it's even possible to encounter duplicate keys from the passed in Map? I think I'm missing some crucial understanding of the issue.

**Comment by mitchswaldman** [comment]
Ah I found the issue. The initial mapping of analyzer providers that comes in will contain a default mapping to the standard analyzer provider. And although I have defined an alias for my custom analyzer to be the default analyzer, this alias isn't picked up until AFTER the initial check to see if "default" is in the provider mapping. When I was debugging ES I found that that my custom analyzer was the first to be iterated over. So the mapping of analyzers was first updated to include my analyzer, then the aliases of my custom analyzer were added to the mapping of analyzers. So after this first iteration, there was a mapping of "default" => CustomAnalyzer. When we finally iterate over the (default => standard analyzer provider) entry, that's where we hit the illegal state exception. So maybe to fix the issue, we move the initial check to see if "default" is missing from the passed in analyzer providers after we iterate through the providers.

**Comment by s1monw** [comment]
what about just never register a default in the AnalysisModule - that should fix the problem

**Comment by mitchswaldman** [comment]
I tried doing that, and the default analyzer snuck in still via the PreBuiltAnalyzers. So I deleted the DEFAULT entry in the PrebuildAnalyzers enumeration, but it still snuck in because the AnalysisService does a check for it in the providers but doesn't look at the possible aliases. So I finally got around it by moving that check after the for loop. Because by then, all the aliases have been registered.

**Comment by s1monw** [comment]
let's look at some code, can you open a PR?

**Comment by mitchswaldman** [comment]
Sure!

**Comment by s1monw** [comment]
I added a testcase to master just to make sure it actually doesn't work but it does. I can reproduce your issue if I alias the default analyzer twice in the mapping, I wonder if you have an issue in your mapping that 5.x detects? Se my test here: https://github.com/elastic/elasticsearch/commit/1cb1373722b97a23f7f551e9827bbedea97b1424

**Comment by mitchswaldman** [comment]
Hmm interesting. I don't think there's anything wrong with my mapping. Here's the settings portion:
"settings": {
    "index.analysis.analyzer.fairhairIndexAnalyzerv3.alias": "default",
    "index.analysis.analyzer.fairhairIndexAnalyzerv3.type": "fairhair-index-analyzer",
    "index.analysis.analyzer.fairhairTokenizingAnalyzer.type": "fairhair-tokenizing-analyzer",
    "index.analysis.analyzer.fairhairTokenizingAnalyzer.alias": "default_search"
  }

**Comment by s1monw** [comment]
wait, what is `fairhair-index-analyzer` and where is is defined?

**Comment by mitchswaldman** [comment]
It's a custom analyzer defined in a plugin. It's registered with the AnalysisModule via the `onModule(AnalysisModule module)` method:

` module.registerAnalyzer(INDEX_ANALYZER_NAME, FairhairIndexAnalyzerProvider::new);
        module.registerAnalyzer(TOKENIZING_ANALYZER_NAME, FairhairTokenizingAnalyzerProvider::new);`

**Comment by s1monw** [comment]
one more question, do you send settings with the create index request since your setting here come from a template?

**Comment by mitchswaldman** [comment]
The settings are being posted to ES once it's up and running. I'm able to see the template with `GET /_template/document-template.json` before there's any indices in ES. Once I try to post data to the index, that's when the "failed to create index exception" happens

**Comment by s1monw** [comment]
@mitchswaldman can you provide a full re-creation of the issue like with curl statements and the bodies of the templates etc?

**Comment by s1monw** [comment]
@mitchswaldman ping

**Comment by mitchswaldman** [comment]
Oh shoot my bad! Of course. 

1) We are running Elasticsearch in Docker so here are the relevant parts of the Dockerfile for installing elasticsearch.

`Dockerfile:`

===================================================
ENV PATH /usr/share/elasticsearch/bin:$PATH

~~~
ENV ES_JAVA_OPTS="$ES_JAVA_OPTS -Xmx1g -Xms1g"

RUN curl -O https://download.elasticsearch.org/elasticsearch/release/org/elasticsearch/distribution/tar/elasticsearch/$ELASTICSEARCH_VERSION/elasticsearch-$ELASTICSEARCH_VERSION.tar.gz
RUN tar -zxvf elasticsearch-$ELASTICSEARCH_VERSION.tar.gz

RUN useradd elasticsearch

RUN mv -f elasticsearch-$ELASTICSEARCH_VERSION/ /usr/share/elasticsearch

RUN set -ex \
  && for path in \
    /usr/share/elasticsearch/data \
    /usr/share/elasticsearch/logs \
    /usr/share/elasticsearch/config \
    /usr/share/elasticsearch/config/scripts \
  ; do \
    mkdir -p "$path"; \
    chown -R elasticsearch:elasticsearch "$path"; \
  done


COPY config/logging.yml /usr/share/elasticsearch/config/

VOLUME /usr/share/elasticsearch/data

COPY docker-entrypoint.sh /
COPY startup.sh /
COPY startup-wrapper.sh /

ENTRYPOINT ["/docker-entrypoint.sh"]

EXPOSE 9200 9300

CMD ["/startup-wrapper.sh"]


ADD plugins /usr/share/elasticsearch/plugins_source


RUN /usr/share/elasticsearch/bin/elasticsearch-plugin install analysis-icu
RUN /usr/share/elasticsearch/bin/elasticsearch-plugin install analysis-phonetic
RUN /usr/share/elasticsearch/bin/elasticsearch-plugin install x-pack
RUN /usr/share/elasticsearch/bin/elasticsearch-plugin install file:////usr/share/elasticsearch/plugins_source/fairhair-elasticsearch-queries-1.19.0_es20/fairhair-elasticsearch-queries-1.19.0-SNAPSHOT.zip
~~~
===================================================

That last plugin is the custom plugin we wrote which registers `FairhairTokenizer` and the other Fairhair analyzers.

2) On start up is when we post the templates to ES.

`startup-wrapper.sh` (referenced in Dockerfile)

===================================================
~~~
set -m

elasticsearch &

while [ $(curl -s -o /dev/null -w "%{http_code}" localhost:9200) != 200 ]; do echo "Waiting for elasticsearch to be ready before posting templates..."; sleep 5; done

for template in $(ls -1 /usr/share/elasticsearch/config/templates); do
  echo -e "\nPosting template: $template"
  curl -s -XPUT "localhost:9200/_template/$template" -d"@/usr/share/elasticsearch/config/templates/$template"
done

echo -e "\nAttaching to elasticsearch process" 
fg %1
~~~
===================================================

3) Here is the template in question. I've removed many of the mappings since they are somewhat redundant.  
`document-template.json`

===================================================
~~~
{
  "order": 0,
  "settings": {
    "index.analysis.analyzer.fairhairIndexAnalyzerv3.alias": "default",
    "index.analysis.analyzer.fairhairIndexAnalyzerv3.type": "fairhair-index-analyzer",
    "index.analysis.analyzer.fairhairTokenizingAnalyzer.type": "fairhair-tokenizing-analyzer",
    "index.analysis.analyzer.fairhairTokenizingAnalyzer.alias": "default_search"
  },
  "mappings": {
    "docnews": {
      "_source": {
        "enabled": true
      },
      "dynamic": "false",
      "_timestamp": {
        "enabled": true
      },
      "_all": {
        "enabled": false
      },
      "properties": {
        "body": {
          "dynamic": false,
          "properties": {
            "byLine": {
              "dynamic": false,
              "properties": {
                "text": {
                  "type": "text",
                  "store": "no",
                  "index": "analyzed",
                  "norms": "false",
                  "index_options": "positions",
                  "doc_values": "false",
                  "analyzer": "fairhairIndexAnalyzerv3"
                }
              }
            },
            "content": {
              "dynamic": false,
              "properties": {
                "text": {
                  "type": "text",
                  "store": "no",
                  "index": "analyzed",
                  "norms": "false",
                  "index_options": "positions",
                  "doc_values": "false",
                  "analyzer": "fairhairIndexAnalyzerv3"
                }
              }
            },
            "ingress": {
              "dynamic": false,
              "properties": {
                "text": {
                  "type": "text",
                  "store": "no",
                  "index": "analyzed",
                  "norms": "false",
                  "index_options": "positions",
                  "doc_values": "false",
                  "analyzer": "fairhairIndexAnalyzerv3"
                }
              }
            },
            "publishDate": {
              "dynamic": false,
              "properties": {
                "date": {
                  "type": "date",
                  "doc_values": "true"
                }
              }
            },
            "title": {
              "dynamic": false,
              "properties": {
                "text": {
                  "type": "text",
                  "store": "no",
                  "index": "analyzed",
                  "norms": "false",
                  "index_options": "positions",
                  "doc_values": "false",
                  "analyzer": "fairhairIndexAnalyzerv3"
                }
              }
            },
            "contentTags": {
              "type": "keyword",
              "index": "not_analyzed",
              "norms": "false",
              "index_options": "docs",
              "doc_values": "true"
            },
            "links": {
              "dynamic": false,
              "type": "nested",
              "properties": {
                "url": {
                  "type": "keyword",
                  "index": "not_analyzed",
                  "norms": "false",
                  "index_options": "docs",
                  "doc_values": "true"
                }
              }
            }
          }
        }
   },
  "template": "document.*"
}
~~~
===================================================

4) The final step in creating the error is to simply POST some data to an index such as `document.document-20160716`. Upon index creation, ES will fail and throw the aforementioned error. 

Let me know if there's anything I need to clear up or add on. For instance, is there a part in the custom queries plugin you need to see?

**Comment by clintongormley** [comment]
Out of interest, why are you doing this:

    "settings": {
      "index.analysis.analyzer.fairhairIndexAnalyzerv3.alias": "default",
      "index.analysis.analyzer.fairhairIndexAnalyzerv3.type": "fairhair-index-analyzer",
      "index.analysis.analyzer.fairhairTokenizingAnalyzer.type": "fairhair-tokenizing-analyzer",
      "index.analysis.analyzer.fairhairTokenizingAnalyzer.alias": "default_search"
    },

Why not just do this:

    "settings": {
      "analysis": {
        "analyzer": {
          "default": {
            "type": "fairhair-index-analyzer"
          },
          "default_search": {
            "type": "fairhair-tokenizing-analyzer"
          }
        }
      }
    }

**Comment by mitchswaldman** [comment]
This template was given to me. So the settings part just came like that.

**Comment by s1monw** [comment]
@mitchswaldman @clintongormley I opened a fix for this issue in #19506 @mitchswaldman would you be able to test this fix if it solves your issue?

**Comment by s1monw** [comment]
@mitchswaldman on another note, I think you should move to the mapping @clintongormley suggested. The entire alias feature seems pretty esoteric and it has lots of caveats like the late binding it does can cause tricky problems. ie.

```
"index.analysis.analyzer.foobar.alias" : "default"
"index.analysis.analyzer.foobar.type", "german"
"index.analysis.analyzer.foobar_search.alias": "default_search"
"index.analysis.analyzer.foobar_search.type": "default"
```

won't work as you expect, it uses the `standard` analyzer as `default_search` but `german` for indexing which is, err not ideal :) - I think one way or the other we will remove this ability unless I understand it's real purpose at some point :)...
